### PR TITLE
[Spark] Add support for writing V2 Checkpoints

### PR DESCRIPTION
**Pull request by dhruvarya-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Adds support for V2 Checkpoints (described in https://github.com/delta-io/delta/issues/1793). This PR only adds write support. More Unit Tests to follow in an upcoming PR which will add read support.

## How was this patch tested?

A basic UT has been added to CheckpointsSuite. More UTs will be added in the read support PR.

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 1]
```diff
@@ -547,25 +574,273 @@ object Checkpoints extends DeltaLogging {
...
+   *
+   * @return A tuple containing
+   *          1. [[FileStatus]] of the newly created top-level V2Checkpoint.
+   *          2. The sequence of actions that were written to the top-level file.
+   *          3. An instance of the LastCheckpointV2 containing V2-checkpoint related
+   *           metadata which can later be written to LAST_CHECKPOINT
+   *          4. Schema of the newly written top-level file (only for parquet files)
+   */
+  protected[delta] def writeTopLevelV2Checkpoint(
+      v2CheckpointFormatOpt: Option[V2Checkpoint.Format],
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:679]: We expect this function to be called only when `v2CheckpointFormatOpt` is non-empty. So we can keep the argument as `v2CheckpointFormat: V2Checkpoint.Format` and the caller could do a `.get`.

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 2]
```diff
@@ -547,25 +574,273 @@ object Checkpoints extends DeltaLogging {
...
+      sidecarCheckpointFiles: Array[SerializableFileStatus],
+      spark: SparkSession,
+      schema: StructType,
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:682]: Some of the arguments are non-trivial (e.g. schema), could we document the parameters in the method doc?

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 3]
```diff
@@ -547,25 +574,273 @@ object Checkpoints extends DeltaLogging {
...
+      Checkpoints.checkpointSchemaToWriteInLastCheckpointFile(spark, schema)
+    val checkpointMetadata = CheckpointMetadata(
+      version = snapshot.version,
+      sidecarNumActions = rowsWrittenInCheckpointJob,
+      sidecarSizeInBytes = parquetFilesSizeInBytes,
+      numOfAddFiles = snapshot.numOfFiles,
+      sidecarFileSchemaOpt = sidecarFileSchemaOpt)
+    val nonFileActionsToWrite =
+      (checkpointMetadata +: sidecarFilesWritten) ++ snapshot.nonFileActions
+    // By default, create v2 checkpoints in json format with fileactions in sidecar files
```
Comment (prakharjain09) [inline spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:702]: This comment doesn't seem to be at the right place? Here we are not relying on the default - we are relying on variable `v2CheckpointFormat`.

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala [hunk 4]
```diff
@@ -424,6 +428,26 @@ object Checkpoints extends DeltaLogging {
     // log store and decide whether to use rename.
     val useRename = deltaLog.store.isPartialWriteVisible(deltaLog.logPath, hadoopConf)
 
+    val v2CheckpointFormatOpt = {
+      val policy = DeltaConfigs.CHECKPOINT_POLICY.fromMetaData(snapshot.metadata)
+      if (policy.needsV2CheckpointSupport) {
+        assert(CheckpointProvider.isV2CheckpointEnabled(snapshot))
+        val v2Format = spark.conf.getOption(DeltaSQLConf.CHECKPOINT_V2_TOP_LEVEL_FILE_FORMAT.key)
```
[the remark this answers is not in the document]
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:435]: `CHECKPOINT_V2_TOP_LEVEL_FILE_FORMAT` is optional, I believe only getOption would work here?

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala (hunk 3 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:702]: Line 689 is wrong ( val v2CheckpointFormat = v2CheckpointFormatOpt.get). The checkpoint v2 format is optional. I should probably move this comment above line 689 and change line 689 to 
```
val v2CheckpointFormat = v2CheckpointFormatOpt.getOrElse( V2Checkpoint.Format.JSON)

Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:702]: Never mind, I forgot that this logic is already present where before `writeTopLevelV2Checkpoint` is called.

Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala:702]: The getOrElse is already taken care of while initializing v2CheckpointFormatOpt - https://github.com/delta-io/delta/pull/2031/files#diff-ffbd94fa1932ccd2f2238ca3ef7911f36c92bfa5bcaf38e3bf43e9fe012e1625R431 .

Also `v2CheckpointEnabled` is defined using  `v2CheckpointFormatOpt` only. Based on this, v2CheckpointFormatOpt must be non-empty here.

File: spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala]: Checked again, .get works here.

**Review by prakharjain09 — APPROVED** [review]

**Review by tdas — APPROVED** [review]
this is a great addition. LGTM.

**Pull request closed without merging**